cpp/src/parquet/variant.h

```diff
@@ -0,0 +1,228 @@
+  Int16 = 4,
+  /// Equivalent Parquet Type: INT(32, signed)
+  Int32 = 5,
+  /// Equivalent Parquet Type: INT(64, signed)
+  Int64 = 6,
+  /// Equivalent Parquet Type: DOUBLE
+  Double = 7,
+  /// Equivalent Parquet Type: DECIMAL(precision, scale)
+  Decimal4 = 8,
+  /// Equivalent Parquet Type: DECIMAL(precision, scale)
+  Decimal8 = 9,
+  /// Equivalent Parquet Type: DECIMAL(precision, scale)
+  Decimal16 = 10,
+  /// Equivalent Parquet Type: DATE
+  Date = 11,
+  /// Equivalent Parquet Type: TIMESTAMP(isAdjustedToUTC=true, MICROS)
+  Timestamp = 12,
+  /// Equivalent Parquet Type: TIMESTAMP(isAdjustedToUTC=false, MICROS)
+  TimestampNTZ = 13,
+  /// Equivalent Parquet Type: FLOAT
+  Float = 14,
+  /// Equivalent Parquet Type: BINARY
+  Binary = 15,
+  /// Equivalent Parquet Type: STRING
+  String = 16,
+  /// Equivalent Parquet Type: TIME(isAdjustedToUTC=false, MICROS)
+  TimeNTZ = 17,
+  /// Equivalent Parquet Type: TIMESTAMP(isAdjustedToUTC=true, NANOS)
+  TimestampTZ = 18,  // Assuming TZ stands for TimeZone, and follows the document's
+                     // 'timestamp with time zone'
+  /// Equivalent Parquet Type: TIMESTAMP(isAdjustedToUTC=false, NANOS)
+  TimestampNTZNanos = 19,  // Differentiating from TimestampNTZ (MICROS)
+  /// Equivalent Parquet Type: UUID
+  Uuid = 20
+};
+
+std::string variantPrimitiveTypeToString(VariantPrimitiveType type);
+
+/// VariantType is from basic type and primitive type.
+enum class VariantType {
+  OBJECT,
+  ARRAY,
+  VARIANT_NULL,
+  BOOLEAN,
+  BYTE,
+  SHORT,
+  INT,
+  LONG,
+  STRING,
+  DOUBLE,
+  DECIMAL4,
+  DECIMAL8,
+  DECIMAL16,
+  DATE,
+  TIMESTAMP_TZ,
+  TIMESTAMP_NTZ,
+  FLOAT,
+  BINARY,
+  TIME,
+  TIMESTAMP_NANOS_TZ,
+  TIMESTAMP_NANOS_NTZ,
+  UUID
+};
+
+std::string variantTypeToString(VariantType type);
+
+class VariantMetadata {
+ public:
+  explicit VariantMetadata(std::string_view metadata);
+  /// \brief Get the variant metadata version. Currently, always 1.
+  int8_t version() const;
+  /// \brief Get the metadata key for a given variant field id.
+  std::string_view getMetadataKey(int32_t variantId) const;
+
+ private:
+  bool sortedStrings() const;
+  uint8_t offsetSize() const;
+  uint32_t dictionarySize() const;
+
+ private:
+  std::string_view metadata_;
+};
+
+template <typename DecimalType>
+struct DecimalValue {
+  uint8_t scale;
+  DecimalType value;
+};
+
+struct VariantValue {
+  VariantMetadata metadata;
+  std::string_view value;
+
+  VariantBasicType getBasicType() const;
+  VariantType getType() const;
+  std::string typeDebugString() const;
+
+  /// \defgroup ValueAccessors
+  /// @{
+
+  // Note: Null doesn't need visitor.
+  bool getBool() const;
+  int8_t getInt8() const;
+  int16_t getInt16() const;
+  int32_t getInt32() const;
+  int64_t getInt64() const;
+  /// Include short_string optimization and primitive string type
+  std::string_view getString() const;
+  std::string_view getBinary() const;
+  float getFloat() const;
+  double getDouble() const;
```

Comment:
Currently, `getDouble` only supports read from `getFloat`, which is too strict for. Maybe we can also uses some way to allow `getDouble` get other types